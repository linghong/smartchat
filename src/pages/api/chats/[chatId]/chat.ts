import 'reflect-metadata';
import { NextApiRequest, NextApiResponse } from 'next';
import {
  DataSource,
  EntityNotFoundError,
  ConnectionIsNotSetError,
  EntityMetadataNotFoundError
} from 'typeorm';

import { getAppDataSource, Chat, User, ChatMessage, ChatImage } from '@/src/db';

const handleDeleteRequest = async (
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  // Start a transaction
  await dataSource.transaction(async transactionalEntityManager => {
    // Find the chat
    const chat = await transactionalEntityManager.findOne(Chat, {
      where: { id: chatIdNum },
      relations: ['messages', 'messages.images']
    });

    if (!chat) {
      return res.status(404).json({ message: 'Chat not found' });
    }

    // Delete all images associated with the chat's messages
    if (chat.messages) {
      for (const message of chat.messages) {
        if (message.images) {
          await transactionalEntityManager.delete(ChatImage, {
            messageId: message.id
          });
        }
      }
    }

    // Delete all messages associated with the chat
    await transactionalEntityManager.delete(ChatMessage, { chatId: chat.id });

    // Finally, delete the chat itself
    await transactionalEntityManager.delete(Chat, { id: chat.id });
  });

  res
    .status(200)
    .json({ message: 'Chat and all associated data deleted successfully' });
};

const handlePutRequest = async (
  req: NextApiRequest,
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  const { title } = req.body;
  const chatRepository = dataSource.getRepository(Chat);

  if (!title) {
    return res.status(400).json({ message: 'Title is required' });
  }

  if (typeof title !== 'string' || title.length < 1 || title.length > 255) {
    return res
      .status(400)
      .json({ message: 'Title must be a string between 1 and 255 characters' });
  }

  // First, find the chat
  const chat = await chatRepository.findOne({ where: { id: chatIdNum } });

  if (!chat) {
    return res.status(404).json({ message: 'Chat not found' });
  }

  // Update the chat title
  chat.title = title;
  await chatRepository.save(chat);

  res.status(200).json({ message: 'Chat title updated successfully' });
};

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  // validation
  const { chatId } = req.query;
  const chatIdNum = Number(chatId);
  if (isNaN(chatIdNum)) {
    return res.status(400).json({ message: 'Invalid chat ID' });
  }

  try {
    const dataSource = await getAppDataSource();

    if (!dataSource)
      return res.status(500).json({ message: 'AppDataSource is null' });

    switch (req.method) {
      case 'DELETE':
        await handleDeleteRequest(res, dataSource, chatIdNum);
        break;
      case 'PUT':
        await handlePutRequest(req, res, dataSource, chatIdNum);
        break;
      default:
        res.setHeader('Allow', ['DELETE', 'PUT']);
        res.status(405).end(`Method ${req.method} Not Allowed`);
    }
  } catch (error) {
    console.error('Error during Chat operation', error);
    if (error instanceof EntityNotFoundError) {
      res.status(404).json({ message: 'Chat not found' });
    } else if (error instanceof ConnectionIsNotSetError) {
      res.status(500).json({ message: 'Database connection error' });
    } else if (error instanceof EntityMetadataNotFoundError) {
      res.status(500).json({ message: 'Entity metadata error' });
    } else {
      res.status(500).json({ message: 'Internal server error' });
    }
  }
};

export default handler;
