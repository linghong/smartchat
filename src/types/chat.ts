export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
