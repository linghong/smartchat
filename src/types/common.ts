export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
