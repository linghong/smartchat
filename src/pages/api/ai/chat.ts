import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';

import getClaudeChatCompletion from '@/src/services/llm/claude';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';

import {
  processImageFiles,
  processNonMediaFiles
} from '@/src/utils/processMessageFile';
import { FileData, ImageFile } from '@/src/types/chat';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { basePrompt, namespace, selectedModel, chatHistory, fileSrc } =
    req.body;

  let { question } = req.body;

  if (!question && fileSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = processImageFiles(fileSrc);

  question += await processNonMediaFiles(fileSrc);

  // Replace multiple newlines with a single newline to avoid excessive spacing
  const pretreatedQuestion = question.trim().replace(/\n\s*\n/g, '\n');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
