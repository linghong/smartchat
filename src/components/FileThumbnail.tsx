import React from 'react';

import { FileData } from '@/src/types/chat';

interface FileThumbnailProps {
  fileData: FileData;
}

const FileThumbnail: React.FC<FileThumbnailProps> = ({ fileData }) => {
  return (
    <div className="w-24 h-24 relative group">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 100 100"
        className="w-full h-full"
      >
        <g className="fill-gray-100 stroke-gray-700 stroke-2">
          <path d="M20 10 H70 L80 20 V90 H20 Z" />
          <path d="M70 10 V20 H80" className="fill-none" />
        </g>
        <text
          x="50"
          y="55"
          className="text-xs font-sans fill-gray-700"
          textAnchor="middle"
        >
          File
        </text>
        <text
          x="50"
          y="97"
          className="text-[8px] font-sans fill-gray-700 opacity-0 transition-opacity duration-300 group-hover:opacity-100"
          textAnchor="middle"
        >
          {fileData?.name?.length > 20
            ? `${fileData?.name?.substring(0, 17)}...`
            : fileData?.name}
        </text>
      </svg>
      <div className="absolute bottom-0 left-0 right-0 bg-black bg-opacity-50 text-white text-xs p-1 truncate">
        {fileData?.name}
      </div>
    </div>
  );
};

export default FileThumbnail;
