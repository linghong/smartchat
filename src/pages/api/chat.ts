import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import getClaudeChatCompletion from '@/src/services/llm/claude';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';
import { ImageFile } from '@/src/types/chat';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const {
    basePrompt,
    question,
    namespace,
    selectedModel,
    chatHistory,
    imageSrc
  } = req.body;

  if (!question && imageSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = imageSrc?.map((imgSrc: ImageFile) => ({
    ...imgSrc,
    base64Image: imgSrc.base64Image.split(',')[1]
  }));

  // replacing newlines with spaces
  const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
