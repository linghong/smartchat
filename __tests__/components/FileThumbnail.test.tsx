import React from 'react';
import { render, screen } from '@testing-library/react';
import FileThumbnail from '@/src/components/FileThumbnail';
import { FileData } from '@/src/types/chat';

// Mock the FileData type
const mockFileData: FileData = {
  name: 'test-file.txt',
  size: 1024,
  type: 'text/plain',
  lastModified: new Date('2023-01-01').getTime()
};

describe('FileThumbnail', () => {
  it('renders the file name', () => {
    render(<FileThumbnail fileData={mockFileData} />);
    const fileNameText = screen.getByText('test-file.txt', { selector: 'div' });
    expect(fileNameText).toBeInTheDocument();
  });

  it('renders the "File" text in the SVG', () => {
    render(<FileThumbnail fileData={mockFileData} />);
    expect(screen.getByText('File')).toBeInTheDocument();
  });

  it('truncates long file names in the SVG text', () => {
    const longNameFileData: FileData = {
      ...mockFileData,
      name: 'very-long-file-name-that-should-be-truncated.txt'
    };
    render(<FileThumbnail fileData={longNameFileData} />);
    const truncatedText = screen.getByText('very-long-file-na...');
    expect(truncatedText).toBeInTheDocument();
    expect(truncatedText).toHaveClass('opacity-0', 'group-hover:opacity-100');
  });

  it('does not truncate short file names in the SVG text', () => {
    render(<FileThumbnail fileData={mockFileData} />);
    const fileNameText = screen.getByText('test-file.txt', {
      selector: 'text'
    });
    expect(fileNameText).toBeInTheDocument();
    expect(fileNameText).toHaveClass('opacity-0', 'group-hover:opacity-100');
  });

  it('renders the file name in the bottom overlay', () => {
    render(<FileThumbnail fileData={mockFileData} />);
    const overlay = screen.getByText('test-file.txt', { selector: 'div' });
    expect(overlay).toBeInTheDocument();
    expect(overlay).toHaveClass(
      'absolute',
      'bottom-0',
      'left-0',
      'right-0',
      'bg-black',
      'bg-opacity-50',
      'text-white',
      'text-xs',
      'p-1',
      'truncate'
    );
  });

  it('applies correct classes to the main container', () => {
    const { container } = render(<FileThumbnail fileData={mockFileData} />);
    const mainDiv = container.firstChild as HTMLElement;
    expect(mainDiv).toHaveClass('w-24', 'h-24', 'relative', 'group');
  });

  it('renders an SVG with correct attributes', () => {
    const { container } = render(<FileThumbnail fileData={mockFileData} />);
    const svg = container.querySelector('svg');
    expect(svg).toBeInTheDocument();
    expect(svg).toHaveAttribute('xmlns', 'http://www.w3.org/2000/svg');
    expect(svg).toHaveAttribute('viewBox', '0 0 100 100');
    expect(svg).toHaveClass('w-full', 'h-full');
  });

  it('applies correct styles to SVG paths', () => {
    const { container } = render(<FileThumbnail fileData={mockFileData} />);
    const paths = container.querySelectorAll('path');
    expect(paths[0]).toHaveAttribute('d', 'M20 10 H70 L80 20 V90 H20 Z');
    expect(paths[1]).toHaveAttribute('d', 'M70 10 V20 H80');
    expect(paths[1]).toHaveClass('fill-none');
  });

  it('renders correctly', () => {
    const { container } = render(<FileThumbnail fileData={mockFileData} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
