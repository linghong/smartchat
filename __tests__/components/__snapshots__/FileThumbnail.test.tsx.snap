// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileThumbnail renders correctly 1`] = `
<div
  class="w-24 h-24 relative group"
>
  <svg
    class="w-full h-full"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      class="fill-gray-100 stroke-gray-700 stroke-2"
    >
      <path
        d="M20 10 H70 L80 20 V90 H20 Z"
      />
      <path
        class="fill-none"
        d="M70 10 V20 H80"
      />
    </g>
    <text
      class="text-xs font-sans fill-gray-700"
      text-anchor="middle"
      x="50"
      y="55"
    >
      File
    </text>
    <text
      class="text-[8px] font-sans fill-gray-700 opacity-0 transition-opacity duration-300 group-hover:opacity-100"
      text-anchor="middle"
      x="50"
      y="97"
    >
      test-file.txt
    </text>
  </svg>
  <div
    class="absolute bottom-0 left-0 right-0 bg-black bg-opacity-50 text-white text-xs p-1 truncate"
  >
    test-file.txt
  </div>
</div>
`;
